src/main/java/graphql/language/AstMultiZipper.java

```diff
@@ -0,0 +1,111 @@
+package graphql.language;
+
+import graphql.PublicApi;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static graphql.Assert.assertTrue;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+@PublicApi
+public class AstMultiZipper {
+    private final Node commonRoot;
+    private final List<AstZipper> zippers;
+
+    public AstMultiZipper(Node commonRoot, List<AstZipper> zippers) {
+        this.commonRoot = commonRoot;
+        this.zippers = new ArrayList<>(zippers);
+    }
+
+    public Node toRootNode() {
+        if (zippers.size() == 0) {
+            return commonRoot;
+        }
+
+        List<AstZipper> curZippers = new ArrayList<>(zippers);
+        while (curZippers.size() > 1) {
+
+            List<AstZipper> deepestZippers = getDeepestZippers(curZippers);
+            Map<Node, List<AstZipper>> sameParent = zipperWithSameParent(deepestZippers);
+
+            List<AstZipper> newZippers = new ArrayList<>();
+            for (Map.Entry<Node, List<AstZipper>> entry : sameParent.entrySet()) {
+                AstZipper newZipper = moveUp(entry.getKey(), entry.getValue());
+                newZippers.add(newZipper);
+            }
+            curZippers.removeAll(deepestZippers);
+            curZippers.addAll(newZippers);
+        }
+        assertTrue(curZippers.size() == 1, "illegal state");
```

Comment:
better assertion message - eg state the precondition better or have no message

Reply:
done